### DISCUSSION TITLE
How to make heading numberings behave differently using Beamer and PDF outputs?

### DISCUSSION THREAD

**Discussion by darthlite** [quarto-dev/quarto-cli#6117] [Q&A] [answered]
### Description

When outputting to Beamer, I have the `slide-level: 3` to allow for sections, subsections, and then slides.

However, in some cases, I go directly from a level one heading to a level three heading in Beamer (i.e., I sometimes have no subsections). But then, when I output the same qmd file to a pdf, the section numbers for those subsubsections/slides then have a 0 in the missing subsection position, e.g., 1.0.1. 

While not the end of the world, it is mildly annoying me and I'm hoping to find a way to essentially skip zeros in subsection numbers when outputting to a pdf to keep my Beamer output intact. Or, put differently, promote subsubsections to subsections if no subsection exists in that section.

Is there a way of easily doing this?

**Comment by mcanouil** [comment]
You can defined specific option in you frontmatter or `_quarto.yml` file.
```yaml
format:
  pdf:
    title: "My Titlte for PDF"
  beamer:
    title: "My Title for Beamer"
```

See the corresponding "reference" pages:
- <https://quarto.org/docs/reference/formats/pdf.html>.
- <https://quarto.org/docs/reference/formats/presentations/beamer.html>.

you might be interested in the Lua filter approach to change section level, see:
- <https://github.com/quarto-dev/quarto-cli/discussions/5853#discussioncomment-6126603>.

---

Could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Comment by darthlite** [comment]
Here is a minimal working example.

```` qmd
---
format: 
  beamer:
    slide-level: 3
  pdf: 
    number-sections: true
---

# Section 1

### Subsubsection (New slide in Beamer)
Jumping straight from h1 to h3.

# Section 2

## Subsection

### Subsubsection
````

PDF output contains a 0 in the subsection position for the first section since no subsection exists:
![Screenshot 2023-07-06 at 3 56 39 PM](https://github.com/quarto-dev/quarto-cli/assets/116678368/1c3ae648-e8e4-44c1-a541-8e9bc798ad7f)

Beamer output is as expected with h3 denoting slide level which is my desired level since I sometimes have sections and subsections in the presentation.

If there is a way to detect when no subsection exists and get rid of that 0 in the pdf and elevate h3 to h2 whenever h2 is missing, that would be nice.

I'm not experienced at all with Lua and had a look at the reply you linked to, but it seemed that it changes all instances of a heading level to another one if I understood correctly(?). In my case, I would like it to be conditioned on if a second level is missing.

**Reply by cderv** [reply]
> I'm not experienced at all with Lua and had a look at the reply you linked to, but it seemed that it changes all instances of a heading level to another one if I understood correctly(?). In my case, I would like it to be conditioned on if a second level is missing.

Do you mean you would like your `h3` level to become `h2` level when `format: pdf` ?

**Reply by darthlite** [reply]
Yes, exactly, but only if there is no h2 for that h3. In the example, the second group of headings are complete (h1, h2, h3), so no change should be made there.

**Reply by cscheid** [reply]
Unfortunately that's not really easy to do in the current crossref system, because that numbering is determined by LaTeX, and not us. I think a better solution would be for you to write a Lua filter that detects the different formats and modifies the heading level explicitly.

**Reply by cderv** [reply, marked as answer]
yes Lua filter is the initial proposal by @mcanouil above. 

Though I recognized it is not an easy one - here is a small working version.

````lua
-- We need to traverse the document from top to bottom for this filter
traverse = 'topdown'

local h2_exists = false

function Header(el)
  if el.level == 1 then
    -- if h1 then we reset the value
    quarto.log.output("H1 found")
    h2_exists = false
  elseif el.level == 2 then
    -- is there an h2 ?
    quarto.log.output("H2 found")
    h2_exists = true;
  elseif el.level == 3 and not h2_exists then
    -- if h3 without h2 then move h3 as h2
    quarto.log.output("H3 found before H2")
    el.level = 2
  end
return el
end
````

I added comment and `quarto.log.output()` to log while running so that you see how it behave. But these lines can be removed. 

You can set use this Lua filter on the `format: pdf` part

````markdown
---
format: 
  beamer:
    slide-level: 3
  pdf: 
    number-sections: true
    filters: 
      - fix-headers.lua
---

# Section 1

### Subsubsection (New slide in Beamer)

Jumping straight from h1 to h3.

# Section 2

## Subsection

### Subsubsection

````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/1b5d7ff9-deda-46ee-b175-f5cf2d499a84)

**Reply by mcanouil** [reply]
FYI, Quarto does include a guide to develop Lua filter and use the "native" format to actually see the standardised format underneath everything, see <https://quarto.org/docs/extensions/lua.html>.

**Comment by darthlite** [comment]
Thank you all very much. Can confirm the filter shared by @cderv does the job. Thank you for all the details about lua, as well.
